### ISSUE TITLE
[MNG-7110] Different behavior of extensions

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8348]
**[Mirko Friedenhagen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mfriedenhagen)** opened **[MNG-7110](https://issues.apache.org/jira/browse/MNG-7110?redirect=false)** and commented

* In my POM I configured the net.rumati.maven.plugins:velocity-maven-plugin which should render a HTML file.
* The parent of the project has a global extension in the build section.
* The jar of the extension does include a velocity which is picked up quite fine by the velocity-maven-plugin from
  the classpath by 3.6.3 but when using 4.0.0 I now need to add the extension jar as a dependency to the plugin.

---

**Affects:** 3.0-alpha-1

**Remote Links:**
- [https://github.com/mfriedenhagen/plugin-classpath-maven-3-and-4
](https://github.com/mfriedenhagen/plugin-classpath-maven-3-and-4)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

When comparing the debug output and ignoring the obvious differences there is one that is very interesting:

```
[DEBUG] Created new class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG] Importing foreign packages into class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG]   Imported:  < maven.api
[DEBUG] Populating class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG]   Included: net.oneandone.maven.poms:foss-configs:jar:1.2
[DEBUG]   Included: org.codehaus.plexus:plexus-utils:jar:1.1
[DEBUG] Extension realms for project com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:jar:1.0-SNAPSHOT: [ClassRealm[extension>net.oneandone.maven.poms:foss-configs:1.2, parent: sun.misc.Launcher$AppClassLoader@4e25154f]]
[DEBUG] Created new class realm project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT
[DEBUG] Populating class realm project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT
[DEBUG]   Included: net.oneandone.maven.poms:foss-configs:jar:1.2
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]
[DEBUG] Extension realms for project net.oneandone.maven.poms:foss-parent:pom:1.6.1: [ClassRealm[extension>net.oneandone.maven.poms:foss-configs:1.2, parent: sun.misc.Launcher$AppClassLoader@4e25154f]]
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]
```

```
[DEBUG] Created new class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG] Importing foreign packages into class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG]   Imported:  < maven.api
[DEBUG] Populating class realm extension>net.oneandone.maven.poms:foss-configs:1.2
[DEBUG]   Included: net.oneandone.maven.poms:foss-configs:jar:1.2
[DEBUG] Extension realms for project com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:jar:1.0-SNAPSHOT: [ClassRealm[extension>net.oneandone.maven.poms:foss-configs:1.2, parent: sun.misc.Launcher$AppClassLoader@4e25154f]]
[DEBUG] Created new class realm project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT
[DEBUG] Populating class realm project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT
[DEBUG]   Included: net.oneandone.maven.poms:foss-configs:jar:1.2
[DEBUG] Looking up lifecycle mappings for packaging jar from ClassRealm[project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]
[DEBUG] Extension realms for project org.codehaus.groovy:groovy-all:pom:2.5.4: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.spockframework:spock-bom:pom:1.2-groovy-2.5: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project net.oneandone.maven.poms:foss-parent:pom:1.6.1: [ClassRealm[extension>net.oneandone.maven.poms:foss-configs:1.2, parent: sun.misc.Launcher$AppClassLoader@4e25154f]]
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[project>com.github.mfriedenhagen:plugin-classpath-maven-3-and-4:1.0-SNAPSHOT, parent: ClassRealm[maven.api, parent: null]]
[DEBUG] Extension realms for project org.codehaus.groovy:groovy-all:pom:2.5.4: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
[DEBUG] Extension realms for project org.spockframework:spock-bom:pom:1.2-groovy-2.5: (none)
[DEBUG] Looking up lifecycle mappings for packaging pom from ClassRealm[plexus.core, parent: null]
```

For some reason the extension dependencies are now handled as projects, which is clearly a bug.
Maybe [gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet) has an idea, I'll need to investigate this.

**Comment by jira-importer** [comment]
**[Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy)** commented

Fixed with [aa90370ff448c68e49ba54bac172f5f7c783292e](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=aa90370ff448c68e49ba54bac172f5f7c783292e) and verified with [c18cf642b539d23996ea00f58fbb4a31714e6fb5](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git;a=commit;h=c18cf642b539d23996ea00f58fbb4a31714e6fb5)
